addons/hr_expense/models/hr_expense_sheet.py

```diff
@@ -464,6 +464,23 @@ def create(self, vals_list):
         sheets.activity_update()
         return sheets
 
+    def write(self, values):
+        res = super().write(values)
+        # Ensures there is no empty expense report in a state different from draft or cancel
+        if 'state' in values and values['state'] or 'expense_line_ids' in values or 'approval_state' in values:
+            for sheet in self:
+                state = sheet.state
+                if sheet.expense_line_ids:
+                    continue
+                if state in {'submit', 'approve', 'post', 'done'} or sheet.approval_state:
```

Comment:
I may be wrong but I think the check on approval_state isn't necessary as `state` is computed based on the approval_state too.
Also, I think the `approval_state` makes the `cancel` state not ok with the current implementation.

```py
    @api.depends('account_move_ids', 'payment_state', 'approval_state')
    def _compute_state(self):
        for sheet in self:
            if sheet.payment_state != 'not_paid':
                sheet.state = 'done'
            elif sheet.account_move_ids:
                sheet.state = 'post'
            elif sheet.approval_state:
                sheet.state = sheet.approval_state
            else:
                sheet.state = 'draft'
```
and the approval state are the following:
```py
    approval_state = fields.Selection(
        selection=[
            ('submit', 'Submitted'),
            ('approve', 'Approved'),
            ('cancel', 'Refused'),
        ],
        copy=False,
    )
```